### ISSUE TITLE
Stripe SDK size is huge

### ISSUE THREAD

**Issue by angelolloqui** [stripe/stripe-ios#1820]
## Summary
App size is very important for our customers, where most of them will download the app over 3G while trying to make a booking in place. Lately, we have observed our app size growing and, together with other factors, we have seen an increase of the size of Stripe with our last update (we migrated from a version of more than 2 years old). Inspecting the binary, you can see how stripe is the biggest library by far in our project (ignoring the swift libraries), even bigger than our whole project code. However, we make very limited used of it, since we mostly interact with the API only, and we do not make use of any of the provided UI. Is there a way where we can import a lighter version of the SDK? this is already very important for us, but it will be crucial soon since we have in roadmap an AppClip where we expect people to pay, and with this binary size it will become impossible.

![Screenshot 2021-06-15 at 08 40 58](https://user-images.githubusercontent.com/1382659/122005297-c1da0e80-cdb5-11eb-8f5d-4b8f1d3ea34e.png)

## Code to reproduce
None

## iOS version
All

## Installation method
```   
pod 'Stripe', '~> 21.6.0'
```

## SDK version
21.6.0

**Comment by davidme-stripe** [comment]
@angelolloqui We're looking into a few options for slimming down the SDK. Just curious: If we had an App Clip SDK, would it be okay if it only supported Apple Pay? Or are you looking to support the card text field and other payment methods as well?

**Comment by angelolloqui** [comment]
I would need to give it a thought with the product team, but In our particular case, we were expecting to support basically all our existing payment methods, so credit cards but iDeal also. For us, if you split the SDK in 2 chunks, with api calls on one side and UI components on the other would be great, since we only use the apis, but I understand it is a hard job to do...

**Comment by davidme-stripe** [comment]
Thanks, that's really helpful feedback! If you're interested, we have an experimental ~1MB SDK with just `STPPaymentCardTextField` and Apple Pay, but it doesn't contain support for iDEAL or 3DS2:
`pod 'Stripe-Min', :git => 'https://github.com/stripe/stripe-ios.git', :branch => 'compact-sdk-21.6.0'`

I'd also recommend generating an [App Size Report in Xcode](https://developer.apple.com/documentation/xcode/reducing-your-app-s-size#Create-the-App-Size-Report) to see the final size for App Store distribution after thinning and bitcode recompilation. The SDK is large, but it shouldn't be anywhere near the ~9MB shown in your screenshot.

**Comment by gogoodboy** [comment]
Hi.
I use the Stripe-Min in my appclip. The  Stripe-Min is the only dependency. I import and init  it at  AppDelegate.Then I archive and find that "App size: 3.3 MB compressed, 9.7 MB uncompressed" . I remove Stripe-Min  and archive again，the size is 407kb. "App size: 168 KB compressed, 407 KB uncompressed". Is tha 9.7 MB the correct size or any step I go wrong.
Thanks.

**Comment by davidme-stripe** [comment]
> Hi.
> I use the Stripe-Min in my appclip. The Stripe-Min is the only dependency. I import and init it at AppDelegate.Then I archive and find that "App size: 3.3 MB compressed, 9.7 MB uncompressed" . I remove Stripe-Min and archive again，the size is 407kb. "App size: 168 KB compressed, 407 KB uncompressed". Is tha 9.7 MB the correct size or any step I go wrong.
> Thanks.

That's very strange, is all that size in the Stripe.framework?

Another idea: Is your App Clip written in Objective-C? It sounds like it might be inadvertently bundling the entire Swift Runtime. This shouldn't happen as long as you set the App Clip to target a minimum version of iOS 12.2 or later.

**Comment by gogoodboy** [comment]
> > Hi.
> > I use the Stripe-Min in my appclip. The Stripe-Min is the only dependency. I import and init it at AppDelegate.Then I archive and find that "App size: 3.3 MB compressed, 9.7 MB uncompressed" . I remove Stripe-Min and archive again，the size is 407kb. "App size: 168 KB compressed, 407 KB uncompressed". Is tha 9.7 MB the correct size or any step I go wrong.
> > Thanks.
> 
> @gogoodboy That's very strange, is all that size in the Stripe.framework?
> 
> Another idea: Is your App Clip written in Objective-C? It sounds like it might be inadvertently bundling the entire Swift Runtime. This shouldn't happen as long as you set the App Clip to target a minimum version of iOS 12.2 or later.

Thanks for your tips. 
It is not the App Clip but the full app that use  some Objective-C SDK and target for iOS 11 or later.
I create a new project and integrate Stripe_Min and the other third-party. The size only has 3.x Mb.
It is the business with the full app now. 
Thanks again.

**Comment by gogoodboy** [comment]
> > > Hi.
> > > I use the Stripe-Min in my appclip. The Stripe-Min is the only dependency. I import and init it at AppDelegate.Then I archive and find that "App size: 3.3 MB compressed, 9.7 MB uncompressed" . I remove Stripe-Min and archive again，the size is 407kb. "App size: 168 KB compressed, 407 KB uncompressed". Is tha 9.7 MB the correct size or any step I go wrong.
> > > Thanks.
> > 
> > 
> > @gogoodboy That's very strange, is all that size in the Stripe.framework?
> > Another idea: Is your App Clip written in Objective-C? It sounds like it might be inadvertently bundling the entire Swift Runtime. This shouldn't happen as long as you set the App Clip to target a minimum version of iOS 12.2 or later.
> 
> Thanks for your tips.
> It is not the App Clip but the full app that use some Objective-C SDK and target for iOS 11 or later.
> I create a new project and integrate Stripe_Min and the other third-party. The size only has 3.x Mb.
> It is the business with the full app now.
> Thanks again.

I upgrade the full app to Swift 5.0, change target to 12.3 or later last night. What the most important part is that do not forget change the setting "platform :ios, '12.3' " in Podfile. And now the  applicp size is ok with  my project .Your tips save my time.

**Comment by rcole34** [comment]
Hey @davidme-stripe, I just came across this thread when looking into possible ways to minimize the size of Stripe (among other efforts to reduce the size of our app as a whole). We're currently only using `STPPaymentCardTextField` (and related `STPPaymentMethodCardParams`, `STPCardParams`, `STPAPIClient.createToken`), so I'm intrigued by the Stripe-Min option, or some other potential options to have different subspecs that might allow us to pick and choose to only use a reduced number of files. We're a little hesitant to use it in its current experimental (and maybe a bit stale) state, though.

I know this is probably a back-burner project that is progressing slowly with plenty of competing priorities, but still wanted to check in and see if you have any updates on if this is still being worked on and what potential timelines might look like. Thank you!

**Comment by davidme-stripe** [comment]
Hi @rcole34, thanks for writing in. Here's a status update:

* We plan to release a fully-supported lightweight module for App Clips next week. This module will only support Apple Pay.
* From there, we'll work on splitting out PaymentSheet, STPCardFormView, and the API bindings from the `Stripe` monolith. We don't have a timeline to share yet, but we're bumping up the priority of it due to the recent increase in SDK size. (Our sample app is currently 2.6MB compressed, which is much larger than we'd like!)

**Comment by davidme-stripe** [comment]
We've added a `StripeApplePay` module to the [latest version of the SDK](https://github.com/stripe/stripe-ios/releases/tag/21.12.0). It adds ~1MB to the uncompressed size of an App Clip.

You can adopt this by following our [Apple Pay integration guide](https://stripe.com/docs/apple-pay). We've also added instructions to help you [migrate an existing STPApplePayContext integration to the new module](https://stripe.com/docs/apple-pay?platform=ios#app-clips).

Sorry for the very long wait! We recognize that SDK size is still a problem — as mentioned above, our next goal is to split [PaymentSheet](https://stripe.com/docs/payments/accept-a-payment) into a separate module.

**Comment by orinmurphy** [comment]
@davidme-stripe thanks for the heads up on the 'Stripe-Min' option, I will check that out.
I really only want to use the STPPaymentCardTextField as a standalone UI component. Im not integrating it with an actual Stripe account.

**Comment by davidme-stripe** [comment]
I don't feel comfortable committing to a specific date yet, but to be transparent we're currently targeting to release it next quarter. (Somewhere in the July–September range.)

**Comment by jeannustre** [comment]
Hi @davidme-stripe ! Could we have progress update on this ? 🙏

**Comment by davidme-stripe** [comment]
Hi all, thanks for your patience. We're about to launch iOS SDK 23.0.0, which will split the Payments SDK into five modules. You can mix and match these to embed only the components you need.
* [StripePayments](https://github.com/stripe/stripe-ios/tree/beta/23.0.0-1/StripePayments) contains the API bindings.
* [StripePaymentsUI](https://github.com/stripe/stripe-ios/tree/beta/23.0.0-1/StripePaymentsUI) includes StripePayments, STPPaymentCardTextField, and other UI elements. 
* [StripeApplePay](https://github.com/stripe/stripe-ios/tree/beta/23.0.0-1/StripeApplePay) contains support for Apple Pay.
* [Stripe](https://github.com/stripe/stripe-ios/tree/beta/23.0.0-1/) includes all of the above packages and our legacy integrations.
* [StripePaymentSheet](https://github.com/stripe/stripe-ios/tree/beta/23.0.0-1/StripePaymentSheet) contains our main PaymentSheet UI.

Here are the App Store sizes for each. (Note that this is the *total* size added by the module, including all dependencies. For example, `StripePaymentsUI` shares many components with `StripeApplePay`, so embedding both would only add an incremental ~100-200kb over embedding `StripePaymentsUI` alone.)

```
+----------------------------+-----------------+-------------------+
| Module                     | Compressed size | Uncompressed size |
+----------------------------+-----------------+-------------------+
|                     Stripe |          2247kb |            5099kb |
|             StripePayments |          1335kb |            3207kb |
|           StripePaymentsUI |          1707kb |            3916kb |
|         StripePaymentSheet |          2748kb |            6254kb |
|             StripeApplePay |           425kb |            1018kb |
+----------------------------+-----------------+-------------------+
```

This is an improvement from v22 (3MB compressed, 6.9MB uncompressed), but it's still bigger than we'd like. We'll continue to explore ways to reduce the size of the SDK.

If you'd like to try it out, point Swift Package Manager or your Podfile at the `beta/23.0.0-1` branch. Thanks!

**Comment by davidme-stripe** [comment]
We released this today as [23.0.0](https://github.com/stripe/stripe-ios/releases/tag/23.0.0). We've also posted a [migration guide](https://stripe.com/docs/mobile/ios/sdk-23-migration?lang-ios=swift) with module size information and tips for migrating to the latest version.
